aten/src/ATen/native/mkldnn/xpu/detail/QMatmul.cpp

```diff
@@ -0,0 +1,272 @@
+    dst_dims = {mb, m, n};
+
+    m1_strides = {m1.stride(0), m1.stride(1), m1.stride(2)};
+    if (m2_trans) {
+      m2_strides = {m2.stride(0), m2.stride(1), m2.stride(2)};
+    } else {
+      m2_strides = {m2.stride(0), m2.stride(2), m2.stride(1)};
+    }
+    dst_strides = {dst.stride(0), dst.stride(1), dst.stride(2)};
+  }
+
+  if (with_bias) {
+    bias_dims = get_onednn_dims(b);
+    bias_dt = get_onednn_dtype(b);
+    bias_strides = get_onednn_strides(b);
+  }
+
+  std::unordered_map<int, dnnl::memory> args;
+
+  dnnl::post_ops po;
+  attr.extract_post_ops(dst, true);
+  bool m1_need_zp = (input_zero_point != 0);
+  bool wgh_is_per_channel = weight_scales.numel() > 1;
+
+  dnnl::matmul matmul_p;
+  dnnl::matmul::primitive_desc matmul_pd;
+
+  m1_md = dnnl::memory::desc(m1_dims, m1_dt, m1_strides);
+  m2_md = dnnl::memory::desc(m2_dims, m2_dt, m2_strides);
+  dst_md = dnnl::memory::desc(dst_dims, dst_dt, dst_strides);
+  dnnl::primitive_attr pattr;
+  pattr.set_post_ops(po);
+  pattr.set_scratchpad_mode(dnnl::scratchpad_mode::user);
+
+  at::Tensor m2_sc;
+  if (!wgh_is_per_channel) {
+    pattr.set_scales_mask(DNNL_ARG_WEIGHTS, 0);
+  } else {
+    pattr.set_scales_mask(DNNL_ARG_WEIGHTS, 1 << 1);
+  }
+
+  at::Tensor m1_sc;
+  dnnl::memory::desc m1_sc_md =
+      dnnl::memory::desc({1}, dnnl::memory::data_type::f32, dnnl::memory::format_tag::x);
+  int mask_ac = 0;
+  pattr.set_scales_mask(DNNL_ARG_SRC, mask_ac);
+  if (m1_need_zp) {
+    pattr.set_zero_points_mask(DNNL_ARG_SRC, mask_ac);
+  }
+
+  if (with_bias) {
+    b_md = dnnl::memory::desc(bias_dims, bias_dt, bias_strides);
+    matmul_pd =
+        dnnl::matmul::primitive_desc(engine, m1_md, m2_md, b_md, dst_md, pattr);
+  } else {
+    matmul_pd = dnnl::matmul::primitive_desc(engine, m1_md, m2_md, dst_md, pattr);
+  }
+
+  matmul_p = dnnl::matmul(matmul_pd);
+
+  m1_usr_md = dnnl::memory::desc(m1_dims, m1_usr_dt, m1_strides);
+  m2_usr_md = dnnl::memory::desc(m2_dims, m2_usr_dt, m2_strides);
+  dst_usr_md = dnnl::memory::desc(dst_dims, dst_usr_dt, dst_strides);
+
+  auto m1_usr_m = make_onednn_memory(m1_usr_md, engine, m1.data_ptr());
+  auto m2_usr_m = make_onednn_memory(m2_usr_md, engine, m2.data_ptr());
+  auto dst_usr_m = make_onednn_memory(dst_usr_md, engine, dst.data_ptr());
+
+  auto expected_m1_md = matmul_pd.src_desc();
+  auto expected_m2_md = matmul_pd.weights_desc();
+  auto expected_dst_md = matmul_pd.dst_desc();
+
+  dnnl::memory m1_m = m1_usr_m, m2_m = m2_usr_m, dst_m = dst_usr_m;
+  at::Tensor m1_, m2_, dst_;
+
+  int scratchpad_size = matmul_pd.scratchpad_desc().get_size();
+  at::Tensor scratchpad_tensor = at::empty(
+      {scratchpad_size}, m1.options().dtype(at::kByte), c10::nullopt);
+  auto scratchpad_memory = make_onednn_memory(
+      matmul_pd.scratchpad_desc(), engine, scratchpad_tensor.data_ptr());
+  args.insert({DNNL_ARG_SCRATCHPAD, scratchpad_memory});
+
+  if (attr.with_binary())
+    attr.construct_post_binary(matmul_pd, args);
```

Comment:
Where to insert the args of binary post ops?  the "std::optional<at::Tensor> other, // extra input for binary-post-op"

Reply:
Yes, the binary is designed for handling the binary tensor. We will use `other` tensor at latter PR #135337. But, I suggest we can keep this logic as this code is safe as we check `attr.with_binary()` before any pos-ops appending. We can view this code is routine design in our mamtul integration codes.